ops/testing.py

```diff
@@ -1746,6 +1787,73 @@ def reboot_count(self) -> int:
         """Number of times the charm has called :meth:`ops.Unit.reboot`."""
         return self._backend._reboot_count
 
+    def run_action(self, action_name: str,
+                   params: Optional[Dict[str, Any]] = None) -> ActionOutput:
+        """Simulates a `juju run` command.
+
+        Validates that no required parameters are missing, and that additional
+        parameters are not provided if that is not permitted. Does not validate
+        the types of the parameters - you can use the
+        `jsonschema <https://github.com/python-jsonschema/jsonschema>`_ package to
+        do this in your tests; for example::
+
+            schema = harness.charm.meta.actions["action-name"].parameters
+            try:
+                jsonschema.validate(instance=params, schema=schema)
+            except jsonschema.ValidationError:
+                # Do something about the invalid params.
+                ...
+            harness.run_action("action-name", params)
+
+        *New in version 2.8*
+
+        Args:
+            action_name: the name of the action to run, as found in ``actions.yaml``.
+            params: override the default parameter values found in ``actions.yaml``.
+                If a parameter is not in ``params``, or ``params`` is ``None``, then
+                the default value from ``actions.yaml`` will be used.
+
+        Raises:
+            ActionFailed: if :meth:`ops.ActionEvent.fail` is called. Note that this will
```

Comment:
I'm sure I'm missing some prior context for this design choice, but it feels odd to me to treat action failure as an exception path.

IMHO an exception should be raised by run_action only if the charm raises an uncaught exception. If the action is expected to fail, and the charm exits gracefully after calling action.fail(), then I find it counter-intuitive to have to catch an exception in the test code.

I'd rather assert that the action has failed without coding defensively.

On the other hand, a positive effect of this choice is that it protects the user from missing an action failure. If the action fails, it will do so loudly and make your test fail, so you have to acknowledge that case explicitly OR fix the charm code so the action won't fail. 

In scenario, the philosophy is: you only get an exception if the charm 'exits nonzero', but action.fail() is not necessarily a charm failure.

Reply:
I can see where you're coming from, and we actually had it the other way to begin with, with an `ActionOutput.success` bool attribute. But we thought it'd be too easy to forget to `assert action_output.success` in charm tests, and the charm action could call `event.fail()` but the test would completely miss that.

That's the main reason. In addition, most tests would require `assert action_output.success` (as presumably the non-failure case is much more common), so it'd also make most tests -- of the success case -- shorter.

You're also right the other way around -- that failing loudly protects the user from missing an action failure.